### PR TITLE
box: allow indexing of fixed-size numeric field types

### LINKED ISSUE #9611: Fixed-size numeric types can't be indexed
The new fixed-size numeric types introduced in #9548 can't be indexed:

```
memtx_space.c:877 E> ER_MODIFY_INDEX: Can't create or modify index 'pk' in space 'test': field type 'uint32' is not supported
```

Reproducer:

```Lua
box.cfg{}
box.schema.create_space('test', {format = {{'a', 'uint32'}}})
box.space.test:create_index('pk')
```

### PR DESCRIPTION
**Pull request by Gumix**
The types added by commit 2e0f9a3c265e ("box: introduce fixed-size numeric field types") can't be indexed. Fix it.

Closes #9611

### CODE REVIEW

File: src/box/memtx_space.c [hunk 1]
```diff
@@ -869,11 +869,12 @@ memtx_space_check_index_def(struct space *space, struct index_def *index_def)
...
 	/* Only HASH and TREE indexes checks parts there */
-	/* Check that there are no ANY, ARRAY, MAP parts */
 	for (uint32_t i = 0; i < key_def->part_count; i++) {
 		struct key_part *part = &key_def->parts[i];
-		if (part->type <= FIELD_TYPE_ANY ||
-		    part->type >= FIELD_TYPE_INTERVAL) {
+		if (part->type == FIELD_TYPE_ANY ||
+		    part->type == FIELD_TYPE_INTERVAL ||
+		    part->type == FIELD_TYPE_ARRAY ||
+		    part->type == FIELD_TYPE_MAP) {
```
Comment (locker) [inline src/box/memtx_space.c:877]: Better add a helper function for this check.

File: src/box/tuple_compare.cc [hunk 2]
```diff
@@ -497,18 +513,30 @@ tuple_compare_field(const char *field_a, const char *field_b,
 {
 	switch (type) {
 	case FIELD_TYPE_UNSIGNED:
+	case FIELD_TYPE_UINT8:
+	case FIELD_TYPE_UINT16:
+	case FIELD_TYPE_UINT32:
+	case FIELD_TYPE_UINT64:
```
Comment (locker) [inline src/box/tuple_compare.cc:519]: I'm not sure it's okay to add so many cases to this switch. Let's ask @alyapunov to be sure.

**Review by locker — CHANGES_REQUESTED** [review]

File: test/engine-luatest/gh_9548_fixed_size_numeric_types_test.lua [hunk 3]
```diff
@@ -320,3 +320,59 @@ g.test_format_change = function(cg)
...
+                  {name = 'u8',  type = 'uint8',    is_nullable = is_nullable},
+                  {name = 'i16', type = 'int16',    is_nullable = is_nullable},
+                  {name = 'u16', type = 'uint16',   is_nullable = is_nullable},
+                  {name = 'i32', type = 'int32',    is_nullable = is_nullable},
+                  {name = 'u32', type = 'uint32',   is_nullable = is_nullable},
+                  {name = 'i64', type = 'int64',    is_nullable = is_nullable},
+                  {name = 'u64', type = 'uint64',   is_nullable = is_nullable},
+                  {name = 'f32', type = 'float32',  is_nullable = is_nullable},
+                  {name = 'f64', type = 'float64',  is_nullable = is_nullable}})
+        s:create_index('i8',  {parts = {{field = 'i8'}}})
```
Comment (locker) [inline test/engine-luatest/gh_9548_fixed_size_numeric_types_test.lua:341]: I think it's also worth checking that the hash index works as well.

File: src/box/memtx_space.c (hunk 1 above)
Comment (Gumix) [reply src/box/memtx_space.c]: Done.

File: test/engine-luatest/gh_9548_fixed_size_numeric_types_test.lua (hunk 3 above)
Comment (Gumix) [reply test/engine-luatest/gh_9548_fixed_size_numeric_types_test.lua]: Done.

File: src/box/tuple_compare.cc (hunk 2 above)
Comment (Gumix) [reply src/box/tuple_compare.cc:519]: Added benchmark results to commit message.

**Review by ochaplashkin — APPROVED** [review]

**Review by locker — APPROVED** [review]

**Pull request merged**